### DISCUSSION TITLE
Elevated Permission for RecordAuth

### DISCUSSION THREAD

**Discussion by munxar** [pocketbase/pocketbase#4753] [Q&A] [answered]
Hello

I have a case where users need to authenticate (**users** collection in this case) but *later*(!) can setup a 2FA method.
2FA is not required for everything, and to prevent horrible UX it must be prosponed only if needed for certain endpoints.

I got 2FA working with custom endpoints, so far so good.

### Example:
User is logged in, but has not entered the 2FA code -> only allow some endpoints
User is logged in, with 2FA entered -> allow very secure endpoints

My question now is: **How can I implement elevated permissions?**

### My idea so far:
Authentication with custom endpoint.
Add a custom endpoint that checks the 2fa code, and if valid exchanges the auth token with something like:
```go
token, err := security.NewJWT(
		jwt.MapClaims{
			"id":           record.Id,
			"type":         tokens.TypeAuthRecord,
			"collectionId": record.Collection().Id,
			"2fa":          true
		},
		(record.TokenKey() + app.Settings().RecordAuthToken.Secret),
		app.Settings().RecordAuthToken.Duration,
	)
```

Add a custom middleware that checks for `"2fa": true` in the token and reject everything else.

Is there something wrong with this, or does someone see a better way of doing this?

Thank you

**Comment by munxar** [comment]
After coding a bit, I came to the conclusion that I will leave the token alone, and try to model the 2FA session as collection with relation to the users collection. By doing this, I can use the API rules mechanics to check if access is granted or not and query the current state from the frontend.

But how to get a identifier for each device? the token itself?

Any thoughts on this approach?

**Reply by ganigeorgiev** [reply]
Technically both options are fine as long as the appropriate checks are in place but it is difficult to say which one would be better without understanding the broader picture or how your two factor verification works.

Note that PocketBase APIs are fully stateless (even the tokens are not stored in the database) and there are no sessions in the traditional sense. A request is considered authenticated if it has a valid `Authorization:TOKEN` header.

If you decide to go with the stored collection state approach you may also need to to perform some cleanup after the "2fa session" has ended/expired.

**Reply by munxar** [reply]
Thanks for the clarification.

I tried to find some parts in pocketbase that are more like what I want to do, and I think the way email verification or password resets work, are quite similar to my needs.
I dumpt the idea of multiple levels of authentication, because it's to complicated to build all endpoints by myself. I will use 2fa (via SMS) allways when logging in.

So I my idea is:

1. use the default users create with email and password
2. prevent loggin in in hook OnRecordAuthRequest until user record meets some criteria (2fa_enabled, phone set etc.)
3. implement a custom auth endpoint /2fa/auth that checks the credentials (email/password) and creates a JWT based on the users collection but doesn't sign them in. possible states could be: needs to setup 2fa or has 2fa enabled (send SMS) etc.
The token is similar to a auth token but short exp and type: "otp_setup" or type: "otp_verify" 
4. implement a endpoint /2fa/verify that consumes a token (JWT) from /2fa/auth and if valid and based on the type, signes them in, or sets some state in the user record (like phone and 2fa_enabled)

I'm not sure if I would put a encrypted version of the code send via SMS in the token, or store it on the user record...

Does this make (more) sense to you? Is there any implementastion that has done this before?
(I found this https://gist.github.com/matevarga/daf2273ddea0644ba6a6e91ae22442e0 but I need a second factor and a phone number setup to work with)

**Reply by ganigeorgiev** [reply, marked as answer]
I'm sorry, I'm not sure that I'll be able to help you as I don't fully understand the described steps and the idea of storing `2fa_enabled` flag in the user record.

I'll stop engage further to avoid causing further confusion, but as a last attempt, based on what I could understand from your comments, one possible example flow could be:

1. Create a middleware that will check whether a token is 2fa verified and attach it to your specific routes:
   ```go
    func Check2FA() echo.MiddlewareFunc {
        return func(next echo.HandlerFunc) echo.HandlerFunc {
            return func(c echo.Context) error {
                authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
                if authRecord == nil {
                    return apis.NewUnauthorizedError("", nil)
                }

                token := c.Request().Header.Get("Authorization")
                if token == "" {
                    return next(c)
                }

                // in this case it is considered "safe" to operate directly with the
                // unverified claims because there is an authRecord loaded already
                // which means that the token was verified by the default auth middleware
                claims, _ := security.ParseUnverifiedJWT(token)
                if cast.ToBool(claims["2fa"]) {
                    return next(c) // 2fa token -> proceed
                }

                // check for 403 on the client-side to determine whether to initiate the 2fa flow
                return apis.NewForbiddenError("The action require 2fa", nil)
            }
        }
    }
   ```
   
2. Register the sms code request and verify actions:
   ```go
   app.OnBeforeServe().Add(func(e *core.ServeEvent) error {        
        e.Router.POST("/2fa/request", func(c echo.Context) error {
            // send sms code to the authenticated user
            // ...

            return c.NoContent(http.StatusNoContent)
        }, apis.RequireRecordAuth())

        e.Router.POST("/2fa/verify", func(c echo.Context) error {
            authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)

            // verify the sms code
            // ...

            // generate a 2fa token
            token, err := security.NewJWT(
                jwt.MapClaims{
                    "id":           authRecord.Id,
                    "type":         tokens.TypeAuthRecord,
                    "collectionId": authRecord.Collection().Id,
                    "2fa":          true,
                },
                (authRecord.TokenKey() + app.Settings().RecordAuthToken.Secret),
                300, // valid for 5 mins
            )
            if err != nil {
                return apis.NewBadRequestError("Failed to generate 2fa token", err)
            }

            return c.JSON(http.StatusOK, map[string]string{
                "token": token,
            })
        }, apis.RequireRecordAuth())

        e.Router.GET("/my-2fa-protected-route", func(c echo.Context) error {
            return c.JSON(http.StatusOK, map[string]any{
                "success": true,
            })
        }, apis.RequireRecordAuth(), Check2FA())

        return nil
    })
   ```
3. (optional) Create client-side SDK helpers:

   ```js
    function request2faCode() {
        return pb.send("/2fa/request", {
            method: "POST"
        })
    }

    async function verify2faCode(code) {
        const result = await pb.send("/2fa/verify", {
            method: "POST",
            body: { code }
        })

        return result.token;
    }

    function my2faProtectedRoute(twoFactorToken) {
        return pb.send("/my-2fa-protected-route", {
            method: "GET",
            headers: {
                Authorization: twoFactorToken,
            }
        })
    }
   ```

**Reply by munxar** [reply]
Thank you very much for taking the time to answer this, but I think my explanations (in the beginning) were to much loud thinking and not explaining what I want.

To put it simple: Everytime you login with email / password, a SMS code is sent that you have to enter and only after that you get signed in. 

I dropped the Initial idea with some routes are behind this 2fa and some are not.
So the title is basically wrong at this point, sorry for the confusion.
